src/build.rs

```diff
@@ -0,0 +1,178 @@
+                    errors += 1;
+                }
+            }
+        }
+    }
+    if errors > 0 {
+        error!("there are untracked .rs files in the repository");
+    }
+}
+
+fn create_branch_if_new(repo: &Repository, name: &str, head: &Reference) {
+    if let Ok(_) = repo.find_branch(name, BranchType::Local) {
+        return;
+    }
+
+    println!("creating branch 'cargo-incremental-build'");
+    let commit = repo.find_commit(head.target().unwrap()).unwrap();
+    if let Err(e) = repo.branch(name, &commit, false) {
+        error!("failed to create branch '{}': {}", name, e);
+    }
+}
+
+fn commit_checkpoint(repo: &Repository) {
+    let author = match Signature::now("cargo-incremental", "none") {
+        Ok(author) => author,
+        Err(e) => error!("failed to create git signature: {}", e),
+    };
+
+    let mut index = match repo.index() {
+        Ok(index) => index,
+        Err(e) => error!("{}", e),
+    };
+
+    let mut pathspecs = Vec::new();
+    pathspecs.push("*");
+    let pathspecs = pathspecs;
+
+    if let Err(e) = index.update_all(pathspecs, None) {
+        error!("{}", e);
+    }
+
+    let updated_tree_oid = match index.write_tree() {
+        Ok(oid) => oid,
+        Err(e) => error!("failed to get oid for updated tree: {}", e),
+    };
+
+    let updated_tree = match repo.find_tree(updated_tree_oid) {
+        Ok(tree) => tree,
+        Err(e) => error!("{}", e),
+    };
+
+    let oid = match repo.refname_to_id("refs/heads/cargo-incremental-build") {
+        Ok(oid) => oid,
+        Err(e) => error!("failed to get oid for cargo-incremental branch: {}", e),
+    };
+
+    let last_commit_incr = match repo.find_commit(oid) {
+        Ok(commit) => commit,
+        Err(e) => error!("failed to get commit: {}", e),
+    };
+
+    let mut parents: Vec<&Commit> = Vec::new();
+    parents.push(&last_commit_incr);
+    let parents = parents;
+
+    let result = repo.commit(Some("HEAD"),
+                             &author,
+                             &author,
+                             "checkpoint",
+                             &updated_tree,
+                             parents.as_slice());
+
+    match result {
+        Ok(oid) => println!("Commit: {:?}", oid),
+        Err(e) => error!("Failed to create commit: {}", e),
+    };
+}
+
+pub fn build(args: &Args) {
+    assert!(args.cmd_build);
+
+    let cargo_toml_pathbuf = Path::new(&args.flag_cargo).canonicalize().unwrap();
+    let cargo_toml_path = cargo_toml_pathbuf.as_path();
+
+    let repo = &match util::open_repo(cargo_toml_path) {
+        Ok(repo) => repo,
+        Err(e) => {
+            error!("failed to find repository containing `{}`: {}",
+                   cargo_toml_path.display(),
+                   e)
+        }
+    };
+
+    let repo_dir = cargo_toml_path.parent().unwrap();
+
+    // Check that there are no are untracked .rs files that might affect the build.
+    check_untracked_rs_files(repo);
```

Comment:
I copied the two functions below. We check for different conditions on status. Maybe we could pass a function and an error message to be printed?

```rust 
pub fn check_clean(repo: &Repository) {
    let statuses = match repo.statuses(None) {
        Ok(s) => s,
        Err(err) => error!("could not load git repository status: {}", err),
    };

    let mut errors = 0;
    let dirty_status = Status::all() - STATUS_IGNORED;
    for status in statuses.iter() {
        if status.status().intersects(dirty_status) {
            let stderr = io::stderr();
            let mut stderr = stderr.lock();
            if let Some(p) = status.path() {
                writeln!(stderr, "file `{}` is dirty", p).unwrap();
            }
            errors += 1;
        }
    }
    if errors > 0 {
        error!("cannot run with a dirty repository; clean it first");
    }
}

fn check_untracked_rs_files(repo: &Repository) {
    let statuses = match repo.statuses(None) {
        Ok(s) => s,
        Err(err) => error!("could not load git repository status: {}", err),
    };

    let mut errors = 0;
    for status in statuses.iter() {
        if status.status().intersects(STATUS_WT_NEW) {
            if let Some(p) = status.path() {
                if p.ends_with("rs") {
                    let stderr = io::stderr();
                    let mut stderr = stderr.lock();
                    writeln!(stderr, "file `{}` is untracked", p).unwrap();
                    errors += 1;
                }
            }
        }
    }
    if errors > 0 {
        error!("there are untracked .rs files in the repository");
    }
}
```